### ISSUE TITLE
Expose a StateChanged event to notify so that the client UI can accuretly display the current connection state.

### ISSUE THREAD

**Issue by okipol88** [dotnet/aspnetcore#49219]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

I am trying to display the current status of the HubConnection in a .NET UI.
The problem is it can transition to different states and in the end I only have access to the underlying actual state.
It would be beneficial to notify about state transitions so that the UI can acuretly reflect the actual connection sate.

### Describe the solution you'd like

Expose a `StateChanged` event that notifies about the underlying connection state changes.

NOTE: This was already done here:
[State Changed Implementation]((https://github.com/SignalR/SignalR/blob/f3600c71f83d8312ad61bced0ca547795734d51e/src/Microsoft.AspNet.SignalR.Client/Connection.cs))

**Comment by mitchdenny** [comment]
We already have ```Closed```, ```Reconnecting```, and ```Reconnected``` events - do these not work for you?

**Comment by mitchdenny** [comment]
You can infer those states. If you haven't called ```StartAsync(...)``` then the state is ```Disconnected```. Once you call ```StartAsync(...)``` but before it is returned the state is ```Connecting``` and once ```StartAsync(...)``` returns the state is connected.

I can see why it might be more convient particularly for driving UI elements to expose an event for this though.

**Comment by okipol88** [comment]
@mitchdenny Yes you can. That also means different people will implement their infer / state sync in a potentially different way. 
Why not have this connection state changed event? Alternatively Connecting, Connected, Disconnected/Closed. Adding info about initial or subsequent or event a retry count to the event should cover the scenarios.

**Comment by mitchdenny** [comment]
Any further input @halter73 or @BrennanConroy? @okipol88 pending feedback from the rest of the team I am going to put this on the backlog. I see this as been a potentially useful change but I want to make sure that there wasn't some reason it was removed given the feature used to be in older versions of SignalR.
